### ISSUE TITLE
Prime-driven compression: benchmark expectations and comparative analysis

### ISSUE THREAD

**Issue by zfifteen** [zfifteen/unified-framework#189]
## Executive summary

- The current implementation cannot losslessly reconstruct data:
  - Positions of bytes are not serialized, so decode concatenates clusters in cluster-id order, destroying the original ordering.
  - Differential encoding has an off-by-one bug that duplicates the first value and drops the last in each cluster.
- Compression performance: no entropy coding is applied; per-cluster payloads are stored almost 1:1. Net “compressed” size is original size plus small constant overhead, so ratios are slightly below 1.00.
- The prime density “enhancement” computed from index-only θ(n,k) is independent of data content; it does not confer compressive power in this implementation.
- Traditional algorithms (gzip/bzip2/lzma) will decisively outperform on repetitive/mixed data, and match/slightly expand on incompressible data—while remaining lossless.

---

## What your benchmark will show (expected)

Given the code paths, here’s what to expect without running anything.

### Size formula

- For M non-empty clusters (M ≤ 5) and data length N:
  - Compressed size ≈ N + 8 + 10·M bytes.
  - So compression ratio \( \approx \frac{N}{N + 8 + 10M} \) which is < 1 for any N > 0.
  - For large N, ratio approaches 1.00 from below; overhead is 18–58 bytes.

### Integrity and length checks

- Decompression returns the concatenation of per-cluster sequences, not the original interleaving. Integrity hash will fail.
- Length will often match (because sums of per-cluster lengths equal N), so “decompression_success” (length-only) will show True, but “integrity_verified” will be False.

### Expected comparative outcomes by data type

| Data type         | Prime-driven (current)                         | gzip                       | bzip2                      | lzma                        |
|---|---|---|---|---|
| Sparse (1–5 KB)   | Ratio ≈ 0.94–0.99; integrity ✗                | 2–5× typical               | 2–6× typical               | 3–10× typical               |
| Repetitive (1–5 KB)| Ratio ≈ 0.94–0.99; integrity ✗               | 5–50× typical              | 5–50× typical              | 10–100× typical             |
| Mixed (1–5 KB)    | Ratio ≈ 0.94–0.99; integrity ✗                | 1.5–5× typical             | 1.5–5× typical             | 2–8× typical                |
| Incompressible (1–5 KB) | Ratio ≈ 0.99; integrity ✗               | ~0.9–1.1× (often slight expansion) | ~0.9–1.1× | ~0.9–1.1×              |

- Timing: Prime-driven adds clustering overhead (GMM) per file; gzip/bzip2/lzma are highly optimized C implementations and will likely be faster per byte, especially for small inputs.

---

## Root causes (and precise fixes)

1) Loss of ordering (positions not serialized)
- Problem: Encoder stores positions_by_cluster but never serializes it. Decoder can’t place bytes back.
- Minimal fix:
  - Serialize positions per cluster using gap encoding (delta of positions) plus a compact integer coding (e.g., variable-byte).
  - On decode: allocate output array of size N, reconstruct cluster byte sequences, then place bytes at recorded positions.

2) Differential encoding off-by-one
- Problem: Using np.diff(..., prepend=first) makes the first diff 0, duplicating the first value and losing the last on reconstruction.
- Fix:
  - Encode: diffs = np.diff(cluster_data.astype(int16))  // length = L-1
  - Store first_value separately.
  - Decode: reconstructed[0] = first_value; for i in 1..L-1: reconstructed[i] = reconstructed[i-1] + diffs[i-1].

3) Clustering is index-only, not data-aware
- Problem: θ(n,k) depends only on index, not the byte values. Clusters don’t align with content redundancy.
- Fix (minimal): augment clustering features with data-local signals, e.g., [byte, delta_prev, n mod m, θ(n,k)], or short n-gram stats.

4) No entropy coding for payload
- Problem: You store per-cluster differences raw (after shifting).
- Fix: apply entropy coding on per-cluster streams (zlib/rANS/arithmetic coding). The value of clustering shows up only when you compress residuals.

5) Integrity check/metadata coupling
- Problem: The benchmark constructs a new compressor for decompression; metadata (especially positions) must be read from the compressed stream, not instance memory.
- Fix: fully self-describe the stream. Don’t rely on in-memory state for decode.

---

## Targeted code patches (concise)

Below are minimal, surgical changes to make the scheme lossless and plausibly compressive.

### A) Correct diffs and include positions

- Replace differential encode/decode with length-1 diffs.
- Add varint helpers and gap-encode positions to keep overhead modest.

```python
# Helpers (compact integer coding)
def varint_encode(nums: np.ndarray) -> bytes:
    out = bytearray()
    for x in nums.astype(np.uint64):
        while True:
            b = x & 0x7F
            x >>= 7
            out.append(b | (0x80 if x else 0))
            if not x: break
    return bytes(out)

def varint_decode(buf: bytes, count: int) -> np.ndarray:
    vals, i = [], 0
    for _ in range(count):
        shift, x = 0, 0
        while True:
            b = buf[i]; i += 1
            x |= (b & 0x7F) << shift
            if (b & 0x80) == 0: break
            shift += 7
        vals.append(x)
    return np.array(vals, dtype=np.int64), i
```

- In _encode_clusters:

```python
cluster_positions = np.where(cluster_mask)[0]
cluster_data = data_array[cluster_mask]
if len(cluster_data) > 0:
    first_val = np.uint8(cluster_data[0])
    diffs = np.diff(cluster_data.astype(np.int16))  # length = L-1
    # Gap-encode positions
    pos_gaps = np.diff(np.insert(cluster_positions, 0, 0)).astype(np.int64)
    pos_bytes = varint_encode(pos_gaps)
    diff_bytes = (np.clip(diffs + 128, 0, 255).astype(np.uint8)).tobytes()

    header = np.array([cluster_id, len(cluster_data), first_val, len(pos_bytes)], dtype=np.uint32)
    payload = pos_bytes + diff_bytes
    encoded_segments.append(header.tobytes() + payload)
```

- In _decode_clusters:

```python
# Parse header (now 16 bytes)
header = np.frombuffer(segment_data[:16], dtype=np.uint32)
cluster_id, length, first_value, pos_len = header
pos_bytes = segment_data[16:16+pos_len]
diff_bytes = segment_data[16+pos_len:]
pos_gaps, consumed = varint_decode(pos_bytes, length+1)  # first gap locates first position relative to 0
positions = np.cumsum(pos_gaps)
diffs = np.frombuffer(diff_bytes, dtype=np.uint8).astype(np.int16) - 128
reconstructed = np.empty(length, dtype=np.uint8)
reconstructed[0] = first_value
if length > 1:
    cumsum = np.cumsum(diffs) + int(first_value)
    reconstructed[1:] = np.clip(cumsum, 0, 255).astype(np.uint8)
segments_data[cluster_id] = (positions, reconstructed)
```

- Reassemble in original order:

```python
out = np.empty(original_length, dtype=np.uint8)
for _, (positions, values) in segments_data.items():
    out[positions] = values
return out.tobytes()
```

This makes decode lossless. Note: serializing positions increases overhead; gap+varint keeps it reasonable when positions form longer runs (consider run-length encoding of gaps equal to 1 to shrink further).

### B) Add entropy coding (quick win)

Even a simple pass improves odds of compression:

```python
import zlib
# after building payload
payload = zlib.compress(payload, level=6)
header = np.array([cluster_id, len(cluster_data), first_val, len(payload)], dtype=np.uint32)
```

On decode, use zlib.decompress first.

### C) Make clustering data-aware (simple feature lift)

Replace:

```python
X = theta_values.reshape(-1,1)
```

With something like:

```python
bytes_arr = np.frombuffer(data, dtype=np.uint8).astype(np.int16)
delta = np.diff(np.insert(bytes_arr, 0, bytes_arr[0])).reshape(-1,1)
theta = theta_values.reshape(-1,1)
pos_mod = (np.arange(len(bytes_arr)) % 16).reshape(-1,1)
X = np.hstack([bytes_arr.reshape(-1,1), delta, theta, pos_mod])
X_scaled = self.scaler.fit_transform(X)
```

Then cluster on X_scaled. Now clusters correlate with actual residual structure.

---

## A redesign that aligns with Zeta Shift

Your real leverage arrives when θ(n,k) isn’t just an index decoration but a reversible transform that amplifies structure before entropy coding.

Two promising directions:

1) θ-permutation transform (BWT-like)
- Idea: Stably sort indices by θ(n,k) (or a k-swept blend), permute data by that order, then apply MTF + entropy coding.
- Store only the permutation (or its inverse) compactly (block-wise plus run-length). If θ-ordering reduces high-order entropy, the net gains outstrip permutation cost.
- This is philosophically Zeta Shift: a relativistic reframe of the observable (indexing) to linearize hidden periodicities.

2) Cluster-as-context entropy coding
- Use GMM cluster labels as adaptive contexts for a range coder over byte deltas.
- Train per-cluster symbol models online; switch context per position. This is essentially a prime-geodesic conditional coder.
- Add k-sweep and cluster-count selection that maximizes held-out cross-entropy reduction, not “enhancement factor.”

Both can be made strictly lossless and compared fairly to gzip/bzip2/lzma.

---

## Comparative analysis (after fixes/redesign)

Assuming you implement either A+B (lossless + entropy) or the redesign:

- Strengths of prime-driven approach
  - If θ-permutation or cluster contexts meaningfully lower entropy for certain data classes (e.g., long-range quasi-periodic signals, modular-structured telemetry, or bitstreams with phase-like drift), you can beat general-purpose codecs on those classes.
  - k-sweep (and model selection) can be tuned per corpus using cross-entropy—not hypothetical “prime density” proxies.
- Weaknesses
  - For arbitrary binary data, general-purpose codecs remain stronger due to decades of engineering and sophisticated transforms (BWT, PPM, LZ77/MA, CM).
  - Any transform that requires storing large permutations or positions pays a bit-budget tax; the transform must reduce entropy enough to justify it.

| Algorithm | Transform | Model | When it wins | Risks |
|---|---|---|---|---|
| Prime-driven (θ-permutation) | θ-sort (+ optional k-sweep) | MTF + rANS/arithmetic | Long-range periodic/modular structure | Permutation side-info cost |
| Prime-driven (cluster-context) | Index/data features → GMM | Per-cluster range coder on deltas | Distinct context-dependent byte stats | Overfitting, context overhead |
| gzip | LZ77 + Huffman | Local repeats | Text, repetitive, structured | None for small sizes (fast, robust) |
| bzip2 | BWT + MTF + Huffman | Runs + contexts | Text/logs/blocks | Slower, memory |
| lzma | LZMA (LZ + range) | Large window, strong | Large repetitive/mixed | CPU-heavy |

---

## Benchmark protocol improvements

- Datasets
  - Include: repetitive, sparse, mixed, incompressible, and a real-world corpus (logs, CSV, binaries).
  - Sizes: 1 KiB, 10 KiB, 100 KiB, 1 MiB, 10 MiB to show scale effects.
  - Multiple seeds for random/incompressible to average variance.

- Metrics per test
  - Compression ratio, encode/decode time, memory use (if possible), integrity (hash), and for your method: model selection time and k chosen.

- Fairness and reproducibility
  - Fix your stream to be self-describing (no in-memory dependencies).
  - Turn on entropy coding; otherwise comparisons are misleading.
  - Report settings for gzip/bzip2/lzma (levels).

- Presentation
  - Tables of results and per-case deltas vs baseline; plot ratio vs size.

---

## Final thoughts

You’re reaching for an invariant that reshapes observables before you measure them—that’s the heart of Zeta Shift. Right now the code gestures at that, but the leverage isn’t connected to the data. Make the transform reversible and the modeling data-aware, then let cross-entropy—not prime-density folklore—decide k and the clustering scheme. Do that, and you’ll see where this truly sings.
